File: quickstart_guides/bit_manipulation/python/sum_two_ints.py
"""
Title: Sum of two integers

Problem:
    Calculate the sum of two integers a and b, but you are not allowed to use
    the operator + and -.

Execution: python sum_two_ints.py
"""
import unittest


def sum_two_ints(a_int: int, b_int: int) -> int:
    """Sum two integers."""
    carry = 0
    remainder = 0
    for i in range(32):
        a = a_int & 1 << i
        b = b_int & 1 << i
        t = a ^ b ^ carry

        remainder |= t

        # Case when both bits are 1.
        if (a & b) != 0:
            carry = 1 << (i + 1)
        # Case when a's or b's bit 1 and carry.
        elif carry != 0 and t == 0:
            carry = 1 << (i + 1)
        else:
            carry = 0

    return remainder


class TestSumTwoInts(unittest.TestCase):
    """Unit test for sum_two_ints."""

    def test_1(self):
        a = 1
        b = 2
        self.assertEqual(sum_two_ints(a, b), 3)

    def test_2(self):
        a = -2
        b = 3
        self.assertEqual(sum_two_ints(a, b), 1)


if __name__ == "__main__":
    unittest.main()

File: quickstart_guides/bit_manipulation/java/SumTwoInts.java
/*
*    Title: Sum two integers.
*
*    Problem:
*        Calculate the sum of two integers a and b, but you are not allowed to
*        use the operator + and -.
*
*    Execution: javac SumTwoIntegers.java && java SumTwoIntegers
*/
import java.util.*;


public class SumTwoInts {
    public static int sumTwoInts(int A, int B) {
        int carry = 0; 
		int r = 0;

		for(int i = 0; i < 32; ++i) {
			int a = (A & 1 << i);
			int b = (B & 1 << i);
			int t = a ^ b ^ carry;

			r |= t;

			// Case when both bits 1.
			if((a & b) != 0) {
				carry = 1 << ( i + 1);
			 // Case when a's or b's bit 1 and carry 1.
			} else if(carry != 0 && t == 0) {
				carry = 1 << ( i + 1);
			} else {
				carry = 0;
			}
		}

		return r;
    }

    public static void main(String[] args) {
        int a_1 = 3; 
        int b_1 = 2;

        int a_2 = -2;
        int b_2 = 3;

        assert sumTwoInts(a_1, b_1) == 3;
        assert sumTwoInts(a_2, b_2) == 1;

        System.out.println("Passed all test cases");
    }
}
